ENH: separate anatomical and functional reports per session for densely sampled dataset

celprov opened the pull request:
enh: beyond a certain number of session for a single subject (4), we separate the anatomical report from the functional and we separate the functional reports per session. For fewer sessions than 4, we keep the original behavior of aggregating all sessions and the anatomical in one report.

enh: find the suitable bootstrapfile within generate_reports rather than before

enh: add session_list argument to generate_reports

Builds on top of #3177

Closes #3080

celprov commented:
@oesteban @effigies I merged #3177 into this PR. Please review it. Thank you!

oesteban commented on fmriprep/reports/core.py:
```diff
@@ -54,10 +78,52 @@ def generate_reports(subject_list, output_dir, run_uuid, config=None, work_dir=N
                 file=str(Path(output_dir) / "logs" / f"report-{run_uuid}-{subject_label}.err"),
             )
 
+        if n_ses >= MAX_SES_AGR:
```
Should this not be an option for the CLI? (with `MAX_SES_AGR` being the default)

celprov replied:
Done in https://github.com/nipreps/fmriprep/pull/3191/commits/87db1c9625263e80246988b767ffe9986b0666fe

celprov replied:
Tested it works

effigies commented on fmriprep/reports/core.py:
```diff
@@ -36,10 +40,27 @@ def generate_reports(subject_list, output_dir, run_uuid, config=None, work_dir=N
+        if bootstrap_file is not None:
+            # If a config file is precised, we do not override it
+            html_report = "report.html"
+        elif n_ses < config.execution.max_ses_agr:
```
```suggestion
        elif n_ses <= config.execution.max_ses_agr:
```

effigies commented on fmriprep/reports/core.py:
```diff
@@ -54,10 +75,52 @@ def generate_reports(subject_list, output_dir, run_uuid, config=None, work_dir=N
                 file=str(Path(output_dir) / "logs" / f"report-{run_uuid}-{subject_label}.err"),
             )
 
+        if n_ses >= config.execution.max_ses_agr:
```
`max` says to me maximum before something different happens, not the point at which something different happens.

```suggestion
        if n_ses > config.execution.max_ses_agr:
```

effigies commented on fmriprep/reports/core.py:
```diff
@@ -54,10 +75,52 @@ def generate_reports(subject_list, output_dir, run_uuid, config=None, work_dir=N
+            session_list = [ses[4:] if ses.startswith("ses-") else ses for ses in session_list]
+
+            for session_label in session_list:
+                bootstrap_file = data.load("reports-spec-func.yml")
+                html_report = ''.join(
+                    [f"sub-{subject_label}", f"_ses-{session_label}", "_func.html"]
+                )
+                entities["session"] = session_label
+
+                robj = Report(
+                    output_dir,
+                    run_uuid,
+                    bootstrap_file=bootstrap_file,
+                    out_filename=html_report,
+                    reportlets_dir=reportlets_dir,
+                    plugins=None,
+                    plugin_meta=None,
+                    metadata=None,
+                    **entities,
+                )
+
+                try:
+                    robj.generate_report()
+                except:
+                    # Store the list of subjects for which report generation failed
+                    errors.append(subject_label)
+                    traceback.print_exception(
+                        *sys.exc_info(),
+                        file=str(
+                            Path(output_dir)
+                            / "logs"
+                            / f"report-{run_uuid}-{subject_label}-func.err"
+                        ),
+                    )
```
This should probably be pulled out into a little function instead of repeated from above. Otherwise we risk getting out of sync.

celprov commented on fmriprep/cli/parser.py:
```diff
@@ -401,6 +401,14 @@ def _slice_time_ref(value, parser):
         "correction. Useful for further Tedana processing post-fMRIPrep.",
     )
+    g_outputs.add_argument(
+        "--max-ses-agr",
```
@effigies in the suggestion above do you envision aggregate-session-reports to still correspond to a number of sessions or is it a flag that turns on or off the aggregation ?

effigies replied:
Number of session still makes sense.

celprov commented:
@effigies @oesteban I added a couple of commits since the last time you reviewed it (thank you for that!)
- I added the boilerplate and the error reporting in report-spec.yml and in report-spec-anat.yml. As such, in the case of separation of the reports per session, the full boilerplate (describing the preprocessing of both anatomical and functional) is added only to the anatomical report to avoid duplicating many times the same information. Similarly, all errors whether it is related to anatomical and functional processing is reported only in the anatomical report in case of report separation.
- I applied Chris' review comment, namely
     - renaming the max-ses-aggr argument
     -  separate the report generation in its own function
- I added unit tests to test the generate_report function
     - It tests whether the expected reports are generated and that no other unexpected files are present.
     - It tests whether the boilerplate, when present, is correctly added inside the report.
     - It tests that crashfiles, when present, are reported inside the html report.
     - It tests the ses-* and sub-* prefix are indeed stripped

oesteban commented on fmriprep/cli/parser.py:
```diff
@@ -401,6 +401,14 @@ def _slice_time_ref(value, parser):
+        type=PositiveInt,
+        default=4,
+        help="Maximum number of sessions aggregated in one subject's visual report. "
+        "If exceeded, visual reports are separated per session.",
```
```suggestion
        "If exceeded, visual reports are split by session.",
```

celprov commented:
I don't know what to do about the "Check fmriprep" test, because if I run `python -m black --color fmriprep ` locally, it does make any changes. Additionally, I tried applying those changes by hand, but it keeps complaining about the same lines requesting additional changes.

oesteban commented on fmriprep/cli/run.py:
```diff
@@ -214,11 +212,17 @@ def main():
         from fmriprep.reports.core import generate_reports
 
         # Generate reports phase
+        session_list = (
+            config.execution.bids_filters["bold"]["session"]
+            if config.execution.bids_filters
+            else None
+        )
```
keys "bold" and/or "session" are not guaranteed to exist at this point. If they don't, you'll get KeyError.

```suggestion
        session_list = config.execution.bids_filters.get("bold", {}).get("session", None)
```

oesteban commented on fmriprep/cli/parser.py:
```diff
@@ -401,6 +401,14 @@ def _slice_time_ref(value, parser):
         "correction. Useful for further Tedana processing post-fMRIPrep.",
     )
+    g_outputs.add_argument(
+        "--aggregate-session-reports",
```
If you want the config file to pick it up automatically, the mapped name should match that in the config file:

```suggestion
        "--aggregate-session-reports",
        dest="aggr_ses_reports",
```

oesteban commented on fmriprep/data/reports-spec-anat.yml:
```diff
@@ -0,0 +1,47 @@
+  reportlets:
+  - bids: {datatype: figures, desc: about, suffix: T1w}
+  - custom: boilerplate
+    path: '{reportlets_dir}/logs'
```
I don't think the logs are written out there for fMRIPrep. Also, I don't think the CITATION.html file is written out into logs:

```suggestion
    path: '{output_dir}'
```

celprov replied:
Yes, the CITATION.html is written inside logs. I had the debug that line to make it work, so I'm sure about this.

oesteban commented on fmriprep/data/reports-spec-anat.yml:
```diff
@@ -0,0 +1,47 @@
+      </p>
+    title: Methods
+  - custom: errors
+    path: '{reportlets_dir}/sub-{subject}/log/{run_uuid}'
```
I don't think crashfiles are written out to the reportlets dir:

```suggestion
    path: '{output_dir}/sub-{subject}/log/{run_uuid}'
```

oesteban commented on fmriprep/data/reports-spec.yml:
```diff
@@ -176,5 +176,23 @@ sections:
   reportlets:
   - bids: {datatype: figures, desc: about, suffix: T1w}
+  - custom: boilerplate
+    path: '{reportlets_dir}/logs'
```
ditto

oesteban commented on fmriprep/data/reports-spec.yml:
```diff
@@ -176,5 +176,23 @@ sections:
+      </p>
+    title: Methods
+  - custom: errors
+    path: '{reportlets_dir}/sub-{subject}/log/{run_uuid}'
```
ditto

oesteban reviewed (commented):
Very quick lookthrough

oesteban commented:
But certainly not under the reportlets dir...

On Tue, Jan 30, 2024 at 3:10 PM celprov ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In fmriprep/data/reports-spec-anat.yml
> <https://github.com/nipreps/fmriprep/pull/3191#discussion_r1471291709>:
>
> > +  - bids: {datatype: figures, space: .*, suffix: T1w, regex_search: True}
> +    caption: Spatial normalization of the T1w image to the <code>{space}</code> template.
> +    description: Results of nonlinear alignment of the T1w reference one or more template
> +      space(s). Hover on the panels with the mouse pointer to transition between both
> +      spaces.
> +    static: false
> +    subtitle: Spatial normalization of the anatomical T1w reference
> +  - bids: {datatype: figures, desc: reconall, suffix: T1w}
> +    caption: Surfaces (white and pial) reconstructed with FreeSurfer (<code>recon-all</code>)
> +      overlaid on the participant's T1w template.
> +    subtitle: Surface reconstruction
> +- name: About
> +  reportlets:
> +  - bids: {datatype: figures, desc: about, suffix: T1w}
> +  - custom: boilerplate
> +    path: '{reportlets_dir}/logs'
>
> Yes, the CITATION.html is written inside logs. I had the debug that line
> to make it work, so I'm sure about this.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nipreps/fmriprep/pull/3191#discussion_r1471291709>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAESDRRGJFB7JCALAMQNZ5LYRD5OHAVCNFSM6AAAAABA6ZGRK6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTQNJRGM4DMNRRGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

effigies commented on fmriprep/data/reports-spec.yml:
```diff
@@ -176,5 +176,23 @@ sections:
   reportlets:
   - bids: {datatype: figures, desc: about, suffix: T1w}
+  - custom: boilerplate
+    path: '{output_dir}/logs'
+    bibfile: ['fmriprep', 'data/boilerplate.bib']
+    caption: |
+      <p>We kindly ask to report results preprocessed with this tool using the following boilerplate.</p>
+      <p class="alert alert-info" role="alert">
+      <strong>Copyright Waiver</strong>.
+      The boilerplate text was automatically generated by <em>NiReports</em> with the
+      express intention that users should copy and paste this text into their manuscripts <em>unchanged</em>.
+      It is released under the
+      <a href="https://creativecommons.org/publicdomain/zero/1.0/" target="_blank">CC0 license</a>.
+      </p>
+    title: Methods
+  - custom: errors
+    path: '{output_dir}/sub-{subject}/log/{run_uuid}'
```
```suggestion
    path: '{out_dir}/logs'
    bibfile: ['fmriprep', 'data/boilerplate.bib']
    caption: |
      <p>We kindly ask to report results preprocessed with this tool using the following boilerplate.</p>
      <p class="alert alert-info" role="alert">
      <strong>Copyright Waiver</strong>.
      The boilerplate text was automatically generated by <em>NiReports</em> with the
      express intention that users should copy and paste this text into their manuscripts <em>unchanged</em>.
      It is released under the
      <a href="https://creativecommons.org/publicdomain/zero/1.0/" target="_blank">CC0 license</a>.
      </p>
    title: Methods
  - custom: errors
    path: '{out_dir}/sub-{subject}/log/{run_uuid}'
```

effigies reviewed (commented):
Looks like nireports uses `out_dir`, not `output_dir`:

https://github.com/nipreps/nireports/blob/0d893ff5cbe75961b34227cffc54dab219fa4a94/nireports/assembler/report.py#L285-L289

effigies commented on fmriprep/cli/run.py:
```diff
@@ -214,11 +212,13 @@ def main():
         from fmriprep.reports.core import generate_reports
 
         # Generate reports phase
+        session_list = config.execution.bids_filters.get("bold", {}).get("session", None)
```
```suggestion
        session_list = config.execution.get().get('bids_filters', {}).get("bold", {}).get("session")
```

effigies commented on fmriprep/cli/workflow.py:
```diff
@@ -86,11 +86,17 @@ def build_workflow(config_file, retval):
     if config.execution.reports_only:
         build_log.log(25, "Running --reports-only on participants %s", ", ".join(subject_list))
+        session_list = (
+            config.execution.bids_filters["bold"]["session"]
```
```suggestion
            config.execution.bids_filters.get("bold", {}).get("session")
```

effigies commented:
Going to go ahead and get this in, re-update the style PR, and we can address any concerns with my last-second hacks in a separate PR.

Thanks for all your effort on this!

celprov commented:
Thank you @effigies for your help finalizing it! I'm happy this one is out of the door :)
